subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java

```diff
@@ -145,6 +149,20 @@ public FileCollection getOutputFiles() {
         }
 
         @Override
+        public Set<File> getPreviousOutputs() {
+            TaskExecution previousExecution = history.getPreviousExecution();
+            if (previousExecution == null || previousExecution.getOutputFilesSnapshot() == null) {
+                return Collections.emptySet();
+            }
+            ImmutableCollection<FileCollectionSnapshot> outputFilesSnapshot = previousExecution.getOutputFilesSnapshot().values();
+            Set<File> outputs = new HashSet<File>();
+            for (FileCollectionSnapshot fileCollectionSnapshot : outputFilesSnapshot) {
+                outputs.addAll(fileCollectionSnapshot.getElements());
+            }
+            return outputs;
+        }
```

Comment:
This is almost exactly the same as `getOutputFiles()`, the only difference being is that we include dirs here, too.

`getOutputFiles()` is called via `TaskOutputsInternal.getPreviousOutputFiles()` that is used solely by `SimpleStaleClassCleaner` to... wait for it!... delete stale files.

We should do a few things:

- do we even still need `StaleClassCleaner`?
- if so, figure out if we could use just one implementation of this method
- if we need to keep both, we should consolidate them into a single method
- the single method should use a simple `Set` instead of a `FileCollection`, as we instantly discard iterate and discard the collection in both cases